### PR TITLE
perf(query): Improve Intersection with UID pack

### PR DESCRIPTION
**Pull request by harshil-goel**
This PR improves the performance of IntersectCompressedWithBin. LDBC09 Query went from 38 second -> 22 second. Benchmarks show up to 80% improvement in performance.

### CODE REVIEW

File: algo/uidlist.go [hunk 1]
```diff
@@ -105,34 +105,43 @@ func IntersectCompressedWithBin(dec *codec.Decoder, q []uint64, o *[]uint64) {
...
-				})
-				if qidx >= len(q) {
-					return
-				}
-				if q[qidx] == u {
-					*o = append(*o, u)
-					qidx++
-				}
-				q = q[qidx:]
+		last_q := q[len(q)-1]
```
Comment (mangalaman93) [inline algo/uidlist.go:108]: We should either use lq everywhere or len(q) everywhere.

File: algo/uidlist.go [hunk 2]
```diff
@@ -105,34 +105,43 @@ func IntersectCompressedWithBin(dec *codec.Decoder, q []uint64, o *[]uint64) {
...
+		uids := make([]uint64, 0)
+		for {
+			n := dec.Uids()
```
Comment (mangalaman93) [inline algo/uidlist.go:111]: choose a better name than `n`. For a second, I thought `n` was number of UIDs.

File: algo/uidlist.go [hunk 3]
```diff
@@ -105,34 +105,43 @@ func IntersectCompressedWithBin(dec *codec.Decoder, q []uint64, o *[]uint64) {
...
+			if len(n) == 0 {
+				break
+			}
+			uids = append(uids, n...)
```
Comment (mangalaman93) [inline algo/uidlist.go:115]: We should avoid creating a big list here, that is not necessary. We could just keep doing intersection with one block at a time.

**Review by mangalaman93 — CHANGES_REQUESTED** [review]
Please add unit tests for these functions too

File: algo/uidlist.go [hunk 4]
```diff
@@ -105,34 +105,45 @@ func IntersectCompressedWithBin(dec *codec.Decoder, q []uint64, o *[]uint64) {
...
-			uids = dec.Next()
+			IntersectWithBin(block_uids, q, o)
+			last_uid := block_uids[len(block_uids)-1]
+			qidx := sort.Search(len(q), func(idx int) bool {
+				return q[idx] >= last_uid
+			})
+			if qidx >= len(q) {
+				return
+			}
+			q = q[qidx:]
```
Comment (mangalaman93) [inline algo/uidlist.go:121]: This is same as what you were doing earlier. The slice will keep increasing. I was hoping to avoid that.

**Review by all-seeing-code — DISMISSED** [review]

File: algo/uidlist.go [hunk 5]
```diff
@@ -105,34 +105,45 @@ func IntersectCompressedWithBin(dec *codec.Decoder, q []uint64, o *[]uint64) {
...
+			dec.Next()
 		}
 		return
 	}
 
+	uids := dec.Seek(q[0], codec.SeekStart)
```
Comment (mangalaman93) [inline algo/uidlist.go:127]: nit: we don't actually need it

File: algo/uidlist.go [hunk 6]
```diff
@@ -105,34 +105,45 @@ func IntersectCompressedWithBin(dec *codec.Decoder, q []uint64, o *[]uint64) {
...
-				if qidx >= len(q) {
-					return
-				}
-				if q[qidx] == u {
-					*o = append(*o, u)
-					qidx++
-				}
-				q = q[qidx:]
+		for {
+			block_uids := dec.Uids()
```
Comment (mangalaman93) [inline algo/uidlist.go:109]: Go doesn't prefer underscore in vars.

**Review by mangalaman93 — CHANGES_REQUESTED** [review]

**Review by mangalaman93 — APPROVED** [review]
Great work @harshil-goel. This should improve our query performance too.

**Review by all-seeing-code — APPROVED** [review]
Lgtm

**Review by joshua-goldstein — APPROVED** [review]
Looks good :+1:

**Pull request merged**